sys/security/mac_do/mac_do.c

```diff
@@ -1264,13 +1316,222 @@ SYSCTL_JAIL_PARAM_SYS_SUBNODE(mac, do, CTLFLAG_RW, "Jail MAC/do parameters");
 SYSCTL_JAIL_PARAM_STRING(_mac_do, rules, CTLFLAG_RW, MAC_RULE_STRING_LEN,
     "Jail MAC/do rules");
 
+static void
+dealloc_conf_osd(void *const value)
+{
+	struct mac_do_conf *conf = value;
+
+	if (conf != NULL) {
+		if (conf->exec_paths_str != NULL) {
+			free(conf->exec_paths_str, M_PATH);
+			conf->exec_paths_str = NULL;
+		}
+		free(conf, M_PATH);
+	}
+}
+
+static struct mac_do_conf *
+alloc_conf(void)
+{
+	struct mac_do_conf *const conf = malloc(sizeof(*conf), M_PATH, M_WAITOK | M_ZERO);
+	
+	memset(conf, 0, sizeof(*conf));
+
+	conf->exec_paths_str = malloc(1, M_PATH, M_WAITOK);
+	conf->exec_paths_str[0] = '\0';
+
+	return conf;
+}
+
+static struct mac_do_conf *
+find_conf(struct prison *pr, struct prison **holder)
+{
+	struct mac_do_conf *conf;
+
+	for (;;) {
+		prison_lock(pr);
+		conf = osd_jail_get(pr, conf_osd_jail_slot);
+		if (conf != NULL) {
+			*holder = pr;
+			return conf;
+		}
+
+		prison_unlock(pr);
+
+		if (pr->pr_parent == NULL) break;
+		pr = pr->pr_parent;
+	}
+
+	*holder = NULL;
+	return NULL;
+}
+
+static struct mac_do_conf *
+find_conf_locked(struct prison *pr)
+{
+	struct prison *cpr = pr;
+	struct mac_do_conf *conf;
+
+	for (;;) {
+		conf = osd_jail_get(cpr, conf_osd_jail_slot);
+		if (conf != NULL) return conf;
+
+		if (cpr->pr_parent == NULL) return NULL;
+		cpr = cpr->pr_parent;
+	}
+}
```

Comment:
FYI, all calls to `find_conf_locked()` should be to `find_conf()` instead (the starting prison is never locked, and all other ones must be anyway).